diff --git a/vpp-manager/Makefile b/vpp-manager/Makefile
index bea80ac..e728448 100644
--- a/vpp-manager/Makefile
+++ b/vpp-manager/Makefile
@@ -72,6 +72,9 @@ rebuild-vpp: vpp-build-env
 vpp: clone-vpp clean-vpp vpp-build-env
 	docker run --rm \
 		-e VPP_MGR_DIR=$(CURDIR) \
+		-e http_proxy=${http_proxy} \
+		-e https_proxy=${https_proxy} \
+		-e no_proxy=${no_proxy} \
 		-v $(CURDIR):$(CURDIR):delegated \
 		--user $$(id -u):$$(id -g) \
 		--network=host \
diff --git a/vpplink/binapi/vpp_clone_current.sh b/vpplink/binapi/vpp_clone_current.sh
index 0b755bc..252fb1d 100755
--- a/vpplink/binapi/vpp_clone_current.sh
+++ b/vpplink/binapi/vpp_clone_current.sh
@@ -109,3 +109,5 @@ git_apply_private 0003-acl-acl-plugin-custom-policies.patch
 git_apply_private 0004-capo-Calico-Policies-plugin.patch
 git_apply_private 0005-memif-fix-socket-name-null-termination.patch
 
+# Apply vpp dsa rx patch
+git_apply_private 0006-dsa-rx.patch
